security/op-node-op-signer-integration.md

```diff
@@ -0,0 +1,145 @@
+
+## Usage of op-signer
+Previously, the only usage of op-signer was for the transaction signing of batcher and proposer. This uses the sender address of the tx and the secret key.
+
+However, for block payload signing, the sender address is not needed - we only need to sign the signingHash with the private key from the KMS. Therefore, we could make `p2p.signer.address` flag optional because it's required for the op-signer to be integrated with services like op-node (see `op-service/signer/cli.go`)
+
+After generating the appropriate key in the KMS, you have to add additional field to the config.yaml of op-signer as follows: 
+```yaml
+auth:
+  - name: op-node.primary.mainnet.prod.oplabs.cloud
+    key: projects/oplabs-prod-mainnet/locations/eur6/keyRings/testop-signer/cryptoKeys/op-node/cryptoKeyVersions/1
+// ... other existing keys
+  - name: sequencer-batcher.primary.mainnet.prod.oplabs.cloud
+    key: projects/oplabs-prod-mainnet/locations/eur6/keyRings/op-signer/cryptoKeys/batcher/cryptoKeyVersions/1
+```
+for each signing request, op-signer will look at the requesting server's domain, and look at this yaml file for a matching config. Then, it will look for the gcp kms key that matches the `key` name.
+
+
+## Security
+
+In such service where op-node and op-signer are located in a separate system and need to communicate, we need secure channel, where both parties can trust each other. For this, we use mutual TLS (mTLS). 
+
+With mTLS, we can prove that both the client and the server are verified with their own respective certificate. This allows op-node to trust the op-signer, and op-signer to sign only the requests from valid op-node sequencers. 
+
+Thankfully, op-signer already has mTLS support, as integrated in op-batcher and op-proposer. 
+
+Here, users can specify
+
+- tls.ca
+- tls.cert
+- tls.key
+
+and the mTLS is handled in op-service/tls. We have to set up the correct CA and so that both the op-signer and op-node can verify the certificates from the TLS handshake. 
+
+There has already been discussions about integrating op-signer with op-node, such as
+
+- https://github.com/ethereum-optimism/optimism/pull/4357#discussion_r1045342439
+- https://github.com/ethereum-optimism/optimism/pull/4669#discussion_r1067409473
+
+# Alternatives Considered
+- When we send the Block Payload from op-node to op-signer, we could send the full block payload arguments like the following: 
+```go
+type BlockPayloadArgs struct {
+	domain [32]byte `json:"domain"`
+	chainId *big.Int `json:"chainId"`
+	payloadHash [32]byte `json:"payloadHash"`
+}
+```
+This means we are sending more data to the op-signer, but may be more secure since the op-signer controls all of the logic for signing the payload. 
```

Comment:
I think `op-signer` being configured with expected parameters and checking the RPC method arguments match is a better trade-off. That way you can detect misuse of the API, rather than producing a signature for the assumed domain and finding it's invalid much later when first doing a verification of the produced signature.

Perhaps the RPC should also include the "sender address", so we can assert that matches the expected address. And then it can be used as an intent, in case this same op-signer RPC server can sign for multiple different p2p signer keys.

Reply:
Thanks for the comment! then I'll modify the design doc and code to send the `domain`, `chainId`, `payloadBytesHash` and `senderAddress`, and add validation for these parameters. Then we could add configurable parameter for chainId and senderAddress when spinning up the op-signer, and validate the incoming rpc requests to match these configurations :)

Reply:
And maybe we should bundle them in a struct, rather than specify them as individual function arguments. This will make it easier to extend the input in the future, by adding optional struct fields.

Reply:
This is updated in the design doc as well as the op-node pr here! https://github.com/ethereum-optimism/optimism/pull/12325/commits/12c4820e35fbba947223748fac13a77e91557e8b

Reply:
Is the purpose of the sender address to verify if the request is valid in a expected restricted environment? I mean, if it has the security purpose, attackers can easily change the address of a request.

Reply:
The address should not be trusted, but it should help specify which account is meant to be used, if there are multiple options. The eth signing API supports a from-address, so it makes sense to do the same here.